core/bubbles/mini_workspace_bubble.ts

```diff
@@ -0,0 +1,199 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {Abstract as AbstractEvent} from '../events/events_abstract.js';
+import type {BlocklyOptions} from '../blockly_options.js';
+import {Bubble} from './bubble.js';
+import type {Coordinate} from '../utils/coordinate.js';
+import * as dom from '../utils/dom.js';
+import {Options} from '../options.js';
+import {Svg} from '../utils/svg.js';
+import type {Rect} from '../utils/rect.js';
+import {Size} from '../utils/size.js';
+import {WorkspaceSvg} from '../workspace_svg.js';
+
+export class MiniWorkspaceBubble extends Bubble {
+  private svgDialog: SVGElement;
+  private miniWorkspace: WorkspaceSvg;
+  private autoLayout = true;
+
+  constructor(
+    workspaceOptions: BlocklyOptions,
+    protected readonly workspace: WorkspaceSvg,
+    protected anchor: Coordinate,
+    protected ownerRect?: Rect
+  ) {
+    super(workspace, anchor, ownerRect);
+    const options = new Options(workspaceOptions);
+    this.validateWorkspaceOptions(options);
+
+    this.svgDialog = dom.createSvgElement(
+      Svg.SVG,
+      {
+        'x': Bubble.BORDER_WIDTH,
+        'y': Bubble.BORDER_WIDTH,
+      },
+      this.contentContainer
+    );
+    workspaceOptions.parentWorkspace = this.workspace;
+    this.miniWorkspace = this.newWorkspaceSvg(new Options(workspaceOptions));
+    const background = this.miniWorkspace.createDom('blocklyMutatorBackground');
+    this.svgDialog.appendChild(background);
+    if (options.languageTree) {
+      background.insertBefore(
+        this.miniWorkspace.addFlyout(Svg.G),
+        this.miniWorkspace.getCanvas()
+      );
+      const flyout = this.miniWorkspace.getFlyout();
+      flyout?.init(this.miniWorkspace);
+      flyout?.show(options.languageTree);
+    }
+
+    this.miniWorkspace.addChangeListener(this.updateBubbleSize.bind(this));
+    this.miniWorkspace
+      .getFlyout()
+      ?.getWorkspace()
+      ?.addChangeListener(this.updateBubbleSize.bind(this));
+    this.updateBubbleSize();
+  }
+
+  dispose() {
+    this.miniWorkspace.dispose();
+    super.dispose();
+  }
+
+  /** @internal */
+  getWorkspace(): WorkspaceSvg {
+    return this.miniWorkspace;
+  }
+
+  /** Adds a change listener to the mini workspace. */
+  addWorkspaceChangeListener(listener: (e: AbstractEvent) => void) {
+    this.miniWorkspace.addChangeListener(listener);
+  }
```

Comment:
What is the value of this method?  Why would a developer not just do `myMWB.getWorkspace().addChangeListener(…)`?  (Do we want to be able to monitor the adding of change listeners?  Seems futile given that the developer can just add it directly as noted.)

Reply:
`getWorkspace` is internal because allowing them to arbitrarily mess with the workspace is a large promise that invites a lot of edge cases I don't have time to test. This creates a more minimal API.

Reply:
> `getWorkspace` is internal

Oh, good point.

I'm sort of tempted to suggest having `getWorkspace` be public but return a `Proxy` that strictly limits what kind of messing they can do, but I will refrain.  (You can tell that I have spent too much time thinking about adversarial programming in JavaScript for Code City.)

Question, though: if they can add a listener, can the listener obtain a reference to the `Workspace` when an event fires?  (See previous parenthetical comment.)

Reply:
That's true, they can obtain a reference to the event via `Blockly.Workspace.getById` from the `workspaceId` property of the event.

Reply:
I think that a proxy would be the most correct solution to this, but I think this is the best compromise for now. Neither the current solution or a proxy would provide complete protection due to `Blockly.Workspace.getById` addressed above (i.e. someone can still always get to the workspace if they want to). But this does limit external developer's access.

If I had more time I would investigate a proxy, but for now I'm going to consider it out of scope.

Reply:
> If I had more time I would investigate a proxy, but for now I'm going to consider it out of scope.

The use of a `Proxy` was not really a serious suggestion (though it would certainly be a way to control access).  I think it would probably be better to invest the time in getting the flyout workspace to support all the regular features (including sub-bubbles!  🫧🫧🫧)